### PR TITLE
[risk=low][RW-3627] Delete project when deleting workspace

### PR DESCRIPTION
**Pull request by jmthibault79**
Description:

~DRAFT.  This will fail until~ https://github.com/broadinstitute/firecloud-orchestration/pull/799 is live

---
**PR checklist**

- [x] This PR meets the Acceptance Criteria in the JIRA story
- [x] The JIRA story has been moved to Dev Review
- [x] This PR includes appropriate unit tests
- [x] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI + API server with `yarn test:local`
- [x] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [x] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [x] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: api/src/bigquerytest/java/org/pmiops/workbench/api/CohortReviewControllerBQTest.java [hunk 1]
```diff
@@ -21,6 +21,7 @@
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
+import org.pmiops.workbench.actionaudit.auditors.BillingProjectAuditor;
```
Comment (jmthibault79) [inline api/src/bigquerytest/java/org/pmiops/workbench/api/CohortReviewControllerBQTest.java:24]: I needed to add this MockBean to a lot of tests after adding it to the WorkspaceServiceImpl constructor.

**Review by petesantos — APPROVED** [review]
LGTM - but I am not the best person for Java PR's

**Comment by jmthibault79** [comment]
OK - I can tag in @jaycarlton as well

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java [hunk 2]
```diff
@@ -270,6 +270,16 @@ public void createAllOfUsBillingProject(String projectName) {
         });
   }
 
+  @Override
+  public void deleteBillingProject(String billingProject) {
+    BillingApi billingApi = billingApiProvider.get();
+    retryHandler.run(
+        (context) -> {
+          billingApi.deleteBillingProject(billingProject);
+          return null;
```
Comment (jaycarlton) [inline api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:279]: nit: you should be able to omit the return statement.

File: api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceServiceImpl.java [hunk 3]
```diff
@@ -306,6 +310,10 @@ public void deleteWorkspace(DbWorkspace dbWorkspace) {
     dbWorkspace.setWorkspaceActiveStatusEnum(WorkspaceActiveStatus.DELETED);
     dbWorkspace = saveWithLastModified(dbWorkspace);
     maybeDeleteRecentWorkspace(dbWorkspace.getWorkspaceId());
+
+    String billingProjectName = dbWorkspace.getWorkspaceNamespace();
```
Comment (jaycarlton) [inline api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceServiceImpl.java:314]: Have you thought through the transactionality of errors here? If we succeed in deleting the workspace, but fail to delete the project, are we happy?

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java [hunk 4]
```diff
@@ -270,6 +270,16 @@ public void createAllOfUsBillingProject(String projectName) {
         });
   }
 
+  @Override
+  public void deleteBillingProject(String billingProject) {
+    BillingApi billingApi = billingApiProvider.get();
+    retryHandler.run(
+        (context) -> {
+          billingApi.deleteBillingProject(billingProject);
```
Comment (jaycarlton) [inline api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:278]: Does this immediately expunge the GCP project, or does it get put into a queue for deletion? There's no wrong answer, as long as everyone understands the behavior.

File: api/src/main/resources/firecloud.yaml [hunk 5]
```diff
@@ -264,6 +264,39 @@ paths:
           - openid
           - email
           - profile
+  /api/user/billing/{projectName}:
+    delete:
+      tags:
+        - Billing
+      summary: delete billing project
+      description: delete billing project
```
Comment (jaycarlton) [inline api/src/main/resources/firecloud.yaml:272]: Please elaborate on this, potentially for folks who haven't worked with the billing system or have forgotten what a billing project is.

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java [hunk 6]
```diff
@@ -270,6 +270,16 @@ public void createAllOfUsBillingProject(String projectName) {
         });
   }
 
+  @Override
+  public void deleteBillingProject(String billingProject) {
```
Comment (jaycarlton) [inline api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:274]: Does this behave the same for user-provided billing projects and free tier billing projects/accounts? Do we need to update spend totals before killing it?

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java (hunk 4 above)
Comment (jaycarlton) [reply api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:278]: Also, is there a ticket for a simple cron to find orphaned billing projects?

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java (hunk 2 above)
Comment (jmthibault79) [reply api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:279]: no - this requires a return value

File: api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceServiceImpl.java (hunk 3 above)
Comment (jmthibault79) [reply api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceServiceImpl.java:314]: Yes - that would be equivalent to the current behavior.  We will always be in a world where we need to go back and clean up orphaned projects (due to local DB wipes) so this is acceptable.

I'll update this to catch-and-log errors

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java (hunk 4 above)
Comment (jmthibault79) [reply api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:278]: Immediate deletion.

We currently have a "Garbage Collection" cron for orphaned billing projects, but we should revisit this process.  I have a plan to do this, but it's not in JIRA yet

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java (hunk 6 above)
Comment (jmthibault79) [reply api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:274]: We will always own the whole lifecycle for billing *projects* - users will bring billing *accounts* in the future which we will assign to the already-existing projects.

File: api/src/test/java/org/pmiops/workbench/workspaces/WorkspaceServiceTest.java [hunk 7]
```diff
@@ -389,4 +390,16 @@ public void enforceFirecloudAclsInRecentWorkspaces() {
     assertThat(recentWorkspaces.size()).isEqualTo(1);
     assertThat(recentWorkspaces.get(0).getWorkspaceId()).isEqualTo(ownedId);
   }
+
+  @Test
+  public void deleteWorkspace() {
```
[the remark this answers is not in the document]
Comment (jmthibault79) [reply api/src/test/java/org/pmiops/workbench/workspaces/WorkspaceServiceTest.java:399]: Mostly this is a passthrough, so there's not much to test.  I added a test for failed project deletion

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java (hunk 6 above)
Comment (jaycarlton) [reply api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:274]: that makes more sense, thanks.

**Review by jaycarlton — APPROVED** [review]
looks good!

**Pull request merged**